android/app/src/main/java/com/zulipmobile/notifications/NotificationHelper.kt

```diff
@@ -65,19 +65,35 @@ fun sizedURL(context: Context, url: URL, dpSize: Float): URL {
     return URL(url, "?$query")
 }
 
-private fun extractName(key: String): String {
-    return key.split(":")[0]
+fun buildNotificationContent(conversations: ByConversationMap, inboxStyle: Notification.InboxStyle, mContext: Context) {
+    for (messages in conversations.values) {
+
+        val messagesByNameMap = buildMessagesByNameMap(messages)
+        val builder = SpannableStringBuilder()
+
+        for (name in messagesByNameMap.keys) {
+            builder.append("$name, ")
```

Comment:
> The messagesByNameMap is actually a LinkedHashMap, so it will maintain the order of the entries. Currently, the order is as the names originally arrived.

Ah, indeed! I should have read the docs on `LinkedHashMap`. :slightly_smiling_face: Now I have -- thanks for pointing that out.


> If, as we discussed before, you want the the last name in the list be the name that sent the last message,, you could remove the key (name) from the map for every new message, and then add it back along with the updated message list.

Yeah. Another algorithm I think would have the same effect as this one: we could iterate through the list in reverse order.

I'm not sure this is the ideal behavior UI-wise, though -- I think we may *also* want the first sender first (with repetition if they're also the last.) Here's how I put it in a comment:

        // TODO ensure latest sender is shown last?  E.g. Gmail-style A, B, ..., A.

That won't be hard to implement either; the main question is the UI one of what behavior we want.

Reply:
Yeah, I wasn't familiar which LinkedHashMap either until I saw it being used for the ByConversationMap class, and had to figure out how it w as storing conversations. 

As for iterating through the list, I'm still thinking in Java and not Kotlin.  In java, to iterate through the list backwards, is messier. You can't use the simple version of the for loop.  You have to do the for(i=0;....) form. Then inside the loop you don't have a nice variable name, you have to do myList.get(i). 

This still might be better than what I suggested above, since it is at least clear that you are iterating through the list backwards.  The reason for removing and then adding back the key on every iteration would not be clear unless it was commented. 

However, Kotlin provides the list.reversed() function., so you can use
for (goodVarName in myList.reversed()) {...}
This is clear and un-messy. It does take more time and memory, but these lists are so small this is a nit.

Reply:
> However, Kotlin provides the list.reversed() function., so you can use
> for (goodVarName in myList.reversed()) {...}
> This is clear and un-messy. It does take more time and memory, but these lists are so small this is a nit.

Yeah, agreed. In the Kotlin docs I also [spotted](https://kotlinlang.org/docs/reference/collection-ordering.html#reverse-order) `asReversed`. I believe that version avoids materializing the reversed list or otherwise costing any extra memory -- the [guide-style docs](https://kotlinlang.org/docs/reference/collection-ordering.html#reverse-order) seem pretty clear about that, though I'd be happier if the API reference for [either](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/reversed.html) [version](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/as-reversed.html) were explicit on that subject. (The Java standard library sets a very high standard in that respect, which Kotlin's doesn't seem to meet.)

If writing in Java I'd probably go for the classic-C-style `for` loop, with a `Foo foo = myList.get(i)` line at the top. Costs a bit of boilerplate, but that seems an OK trade for making the logic clearer. And some boilerplate is to be expected when working in Java :wink:

Reply:
I think the documentation on asReversed is confusing too. The documentation shows examples where asReversed is assigned to a variable, but what actual object it's returning is not clear.

Your right about the C-style loop, especially if you need to use myList.get(i) in more than one place.